### PR TITLE
fix: stream list requests straight to http writer

### PR DESCRIPTION
**Pull request by blakepettersson**
Each list request builds the entire JSON response body in a `bytes.Buffer` before writing it to the client. For large application counts the buffer grows to tens or hundreds of MB per request, and under concurrent load these buffers stack up before GC can reclaim them. Write the list response directly to `http.ResponseWriter` instead of accumulating it in a `bytes.Buffer`. Since `http.ResponseWriter` implements `io.Writer` and is internally buffered by `net/http`, this eliminates the per-request spike entirely — bytes flow straight to the TCP stack as they are produced.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 1]
```diff
@@ -112,6 +128,108 @@ func processApplicationListField(v any, fields map[string]any, exclude bool) (an
...
+				return err
+			}
+		}
+		var data []byte
+		if useFieldFilter {
+			converted, err := processAppFields(&appList.Items[i], fields, exclude)
+			if err != nil {
+				return err
+			}
+			data, err = json.Marshal(converted)
```
Comment (crenshaw-dev) [inline pkg/apiclient/application/forwarder_overwrite.go]: Should we instead use a json encoder to write directly to the writer?

Comment (blakepettersson) [reply pkg/apiclient/application/forwarder_overwrite.go]: Good point, done!

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 2]
```diff
@@ -153,6 +263,50 @@ func init() {
...
+		appList, ok := resp.(*v1alpha1.ApplicationList)
+		if !ok {
+			runtime.ForwardResponseMessage(ctx, mux, marshaler, w, req, resp, opts...)
+			return
+		}
+
+		// Replicate grpc-gateway ForwardResponseMessage header handling.
+		md, ok := runtime.ServerMetadataFromContext(ctx)
+		if ok {
+			for k, vs := range md.HeaderMD {
```
Comment (rumstead) [inline pkg/apiclient/application/forwarder_overwrite.go:337]: Should we do 
```mux.outgoingHeaderMatcher``` to future proof us?

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 3]
```diff
@@ -71,38 +78,47 @@ var appFields = map[string]func(app *v1alpha1.Application) any{
...
+					return nil, fmt.Errorf("field %s is not a map", field)
+				}
+				item = nestedMap
+			}
+		}
+	}
+	return converted, nil
+}
+
 func processApplicationListField(v any, fields map[string]any, exclude bool) (any, error) {
```
Comment (rumstead) [inline pkg/apiclient/application/forwarder_overwrite.go]: I am probably missing something obvious (or my IDE is tricking me) but is this function called outside of tests?

Comment (rumstead) [reply pkg/apiclient/application/forwarder_overwrite.go]: If so, should we move it to a test file?

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 4]
```diff
@@ -112,6 +128,100 @@ func processApplicationListField(v any, fields map[string]any, exclude bool) (an
...
+	if !useFieldFilter {
+		if err := writeInlineTypeMeta(w, &appList.TypeMeta); err != nil {
+			return err
+		}
+	}
+
+	if _, err := w.Write([]byte(`"metadata":`)); err != nil {
+		return err
+	}
+	if err := enc.Encode(appList.ListMeta); err != nil {
```
Comment (rumstead) [inline pkg/apiclient/application/forwarder_overwrite.go:184]: I don't think this is a problem but `enc.Encode` appends a `\n` after each call.

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 5]
```diff
@@ -112,6 +128,100 @@ func processApplicationListField(v any, fields map[string]any, exclude bool) (an
...
+		return err
+	}
+	if _, err := w.Write([]byte(`,"items":[`)); err != nil {
```
Comment (rumstead) [inline pkg/apiclient/application/forwarder_overwrite.go:194]: I believe this will mean that items can no longer be `nil` just `[]` is that ok? Should we add a test for those two cases of list being empty vs nil?

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 6]
```diff
@@ -153,6 +263,50 @@ func init() {
...
+		for _, opt := range opts {
+			if err := opt(ctx, w, resp); err != nil {
+				runtime.HTTPError(ctx, mux, marshaler, w, req, err)
+				return
+			}
+		}
+
+		fields, exclude := parseFieldSelection(req)
+
+		if err := streamApplicationListJSON(w, appList, fields, exclude); err != nil {
```
Comment (rumstead) [inline pkg/apiclient/application/forwarder_overwrite.go:359]: What happens if there is any issue mid stream? Like we cannot marshal something or some network blip? I think the http status code would be 200 and depending on the error the client would get a bunch of unparsable json. Would it make sense to do a `panic(http.ErrAbortHandler)` here?

File: pkg/apiclient/application/forwarder_overwrite.go (hunk 3 above)
Comment (blakepettersson) [reply pkg/apiclient/application/forwarder_overwrite.go]: Yeah this was originally to assert that the new output == old output, but makes more sense to move that to a test file

File: pkg/apiclient/application/forwarder_overwrite.go (hunk 5 above)
Comment (blakepettersson) [reply pkg/apiclient/application/forwarder_overwrite.go:194]: I'll add a guard for `nil` Items

File: pkg/apiclient/application/forwarder_overwrite.go (hunk 6 above)
Comment (blakepettersson) [reply pkg/apiclient/application/forwarder_overwrite.go:359]: Done

File: pkg/apiclient/application/forwarder_overwrite.go (hunk 2 above)
Comment (blakepettersson) [reply pkg/apiclient/application/forwarder_overwrite.go:337]: Sadly `mux.outgoingHeaderMatcher` is unexported 😞

**Comment by crenshaw-dev** [comment]
@blakepettersson I'm a bit hesitant to cherry pick this. Maybe just to 3.4?

**Comment by blakepettersson** [comment]
@crenshaw-dev 3.4 only sounds good to me!

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 7]
```diff
@@ -71,45 +78,138 @@ var appFields = map[string]func(app *v1alpha1.Application) any{
...
+	}
+	if tm.APIVersion != "" {
+		data, err := json.Marshal(tm.APIVersion)
+		if err != nil {
+			return err
+		}
+		if _, err := fmt.Fprintf(w, `"apiVersion":%s,`, data); err != nil {
+			return err
+		}
+	}
```
Comment (gdsoumya) [inline pkg/apiclient/application/forwarder_overwrite.go:237]: Is the json.Marshal needed only to escape the string? If so then I don't think we need it because both kind and api version cannot have escapable characters. We can save some reflection magic and processing by skipping json.Marshal.

**Review by gdsoumya — CHANGES_REQUESTED** [review]

File: pkg/apiclient/application/forwarder_overwrite.go [hunk 8]
```diff
@@ -153,6 +253,51 @@ func init() {
 		}
 		return event.Application.Name, nil
 	})
-	forward_ApplicationService_List_0 = http.UnaryForwarderWithFieldProcessor(processApplicationListField)
+	forward_ApplicationService_List_0 = func(ctx context.Context, mux *runtime.ServeMux, marshaler runtime.Marshaler, w gohttp.ResponseWriter, req *gohttp.Request, resp proto.Message, opts ...func(context.Context, gohttp.ResponseWriter, proto.Message) error) {
+		appList, ok := resp.(*v1alpha1.ApplicationList)
```
Comment (alexmt) [inline pkg/apiclient/application/forwarder_overwrite.go:322]: This forwarder also handlers SSE request: https://github.com/argoproj/pkg/blob/5290b9c2e627a0ab483299f2f9483c0d2e8a9a4a/grpc/http/forwarders.go#L228

I think easiest way to handle is to use old forwarder if `req.Header.Get("Accept") == "text/event-stream"`

**Review by alexmt — DISMISSED** [review]

**Comment by blakepettersson** [comment]
@gdsoumya @alexmt fixed, PTAL

**Review by gdsoumya — APPROVED** [review]

**Comment by blakepettersson** [comment]
@alexmt has your concern been addressed?

I removed the cherry-pick; for now we'll just get this in our Akuity fork.

**Comment by blakepettersson** [comment]
@rumstead @crenshaw-dev did some tweaks to this PR, wdyt?

File: pkg/apiclient/application/forwarder_overwrite_test.go [hunk 9]
```diff
@@ -53,3 +68,538 @@ func TestProcessApplicationListField_SyncOperationMissing(t *testing.T) {
...
+	}))
+	defer ts.Close()
+
+	req, err := http.NewRequestWithContext(t.Context(), http.MethodGet, ts.URL+"/api/v1/applications", http.NoBody)
+	require.NoError(t, err)
+	resp, err := http.DefaultClient.Do(req)
+	if err == nil {
+		resp.Body.Close()
+	}
+	require.ErrorIs(t, err, io.EOF)
```
Comment (rumstead) [inline pkg/apiclient/application/forwarder_overwrite_test.go]: a mega nit but I think this error could also be a connection reset. 

```
require.Error(t, err, "aborted response must not complete cleanly")
```

**Review by rumstead — APPROVED** [review]
left a nit comment on the tests but LGTM

File: pkg/apiclient/application/forwarder_overwrite_test.go (hunk 9 above)
Comment (blakepettersson) [reply pkg/apiclient/application/forwarder_overwrite_test.go]: Done
